### PR TITLE
BC-8211 - Cancel Confirmation on Room Edit/Create

### PR DESCRIPTION
**Pull request by odalys-dataport**
# Short Description

Ask user to confirm that they want to cancel editing or creating a room. Do not open dialog, when the form doesn't show any changes.

## Links to Ticket and related Pull-Requests
BC-8211

## Checklist before merging

- [x] QA: In addition to review, the code has been manually tested (if manual testing is possible)
- [ ] PO: Any deviation from requirements was agreed with Product-Owner / ticket author / support-team
- [ ] DEV: Every new component is implemented having accessibility in mind (e.g. aria-label, role property)

> Notice: Please keep this Pull-Request as a Draft (or add WIP label), until it is ready to be reviewed

### CODE REVIEW

File: src/modules/feature/room/RoomForm.vue [hunk 1]
```diff
@@ -76,14 +84,22 @@ const props = defineProps({
 const emit = defineEmits(["save", "cancel"]);
 const { t } = useI18n();
 
+const { askConfirmation } = useConfirmationDialog();
+
 const roomData = computed(() => props.room);
 
+const onUpdateColor = () => {
+	v$.value.$touch();
```
Comment (wolfganggreschus) [inline src/modules/feature/room/RoomForm.vue:92]: Even it's possible a variable should be declared before using it. First time I saw it I thought it's a global variable.

Comment (wolfganggreschus) [inline src/modules/feature/room/RoomForm.vue:92]: Even though this didn't happen in this PR, naming a variable "v$" is very cryptic and hard to read for people unfamiliar with the code. Suggestion: 
`const validationState = useVuelidate(rules, { roomData }, { $lazy: true, $autoDirty: true });`

File: src/modules/feature/room/RoomForm.vue [hunk 2]
```diff
@@ -106,7 +122,16 @@ const onSave = async () => {
 	emit("save", roomData.value);
 };
 
-const onCancel = () => {
-	emit("cancel");
+const onCancel = async () => {
+	if (!v$.value.$anyDirty) emit("cancel");
```
Comment (wolfganggreschus) [inline src/modules/feature/room/RoomForm.vue:126]: It's a little bit difficult to understand. Suggestion:
`const noChangesMade = !v$.value.$anyDirty;
if (!v$.value.$anyDirty) emit("cancel");`

File: src/modules/feature/room/RoomColorPicker/RoomColorPickerSwatch.unit.ts [hunk 3]
```diff
@@ -5,7 +5,7 @@ import {
 import { mount } from "@vue/test-utils";
 import { ComponentProps } from "vue-component-type-helpers";
 import RoomColorPickerSwatch from "./RoomColorPickerSwatch.vue";
-import { RoomColorEnum } from "./types";
+import { RoomColor } from "@/serverApi/v3";
```
Comment (uidp) [inline src/modules/feature/room/RoomColorPicker/RoomColorPickerSwatch.unit.ts:8]: The type aliases were actually intended to be able to decouple from the api types. Why why are they removed?

File: src/locales/de.ts [hunk 4]
```diff
@@ -156,10 +156,10 @@ export default {
 	"common.words.classes": "Klassen",
 	"common.words.color": "Farbe",
 	"common.words.color.blue": "Blau",
-	"common.words.color.blueGrey": "Blaugrau",
+	"common.words.color.blue-grey": "Blaugrau",
```
[the remark this answers is not in the document]
Comment (odalys-dataport) [reply src/locales/de.ts:159]: Yes, I did this to be able to reuse the color names from the enum. There are some other occurrences where we use kebab case instead as well, but I'm not sure if I am happy with this solution either.

File: src/modules/feature/room/RoomColorPicker/RoomColorPickerSwatch.unit.ts (hunk 3 above)
Comment (odalys-dataport) [reply src/modules/feature/room/RoomColorPicker/RoomColorPickerSwatch.unit.ts:8]: As it's not possible to assign enums to a different variable name, we'll keep the import from api here.

File: src/modules/feature/room/RoomForm.vue (hunk 1 above)
Comment (odalys-dataport) [reply src/modules/feature/room/RoomForm.vue:92]: done :)

Comment (odalys-dataport) [reply src/modules/feature/room/RoomForm.vue:92]: It is a little confusing at first glance, but we decided to stick with `v$` as it's the vuelidate convention.

File: src/modules/feature/room/RoomForm.unit.ts [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+};
+
+describe("@feature-room/RoomForm", () => {
+	const setup = (props: ComponentProps<typeof RoomForm>) => {
+		const wrapper = mount(RoomForm, {
+			global: {
+				plugins: [createTestingVuetify(), createTestingI18n()],
+			},
+			props,
+			attachTo: document.body,
```
Comment (uidp) [inline src/modules/feature/room/RoomForm.unit.ts:31]: Do we need this?

File: src/modules/feature/room/RoomForm.unit.ts [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+	it("should not save invalid room", async () => {
+		const { wrapper } = setup({ room: invalidMockRoom });
+
+		// validation needs to be triggered manually due to code structure
+		await wrapper.vm.v$.$validate();
+
+		await wrapper.find("[type='submit']").trigger("click");
+		await nextTick();
+
+		expect(wrapper.vm.v$.$invalid).toEqual(true);
```
Comment (uidp) [inline src/modules/feature/room/RoomForm.unit.ts:46]: Can we find a better way to watch the behavior of the component other than looking at the internal state? (other than `wrapper.vm`)

**Review by uidp — APPROVED** [review]

File: src/modules/feature/room/RoomForm.unit.ts (hunk 5 above)
Comment (MartinSchuhmacher) [reply src/modules/feature/room/RoomForm.unit.ts:31]: yes, otherwise the state was not observable in the test

**Pull request merged**